File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public boolean isValidBST(TreeNode root) {
        return validRecursive(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    private boolean validRecursive(TreeNode curr, int min, int max)  {
        if (curr == null)   {
            return true;
        }
        if (curr.val <= min || curr.val >= max)   {
            return false;
        }
        return validRecursive(curr.left, min, curr.val) && validRecursive(curr.right, curr.val, max);
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution: # RECURSIVE SOLUTION (BST specific)
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        # Base case:
        # If root is None, or either node is missing (defensive check),
        # we cannot find an LCA
        if not root or not p or not q:
            return None
        
        # If BOTH p and q are smaller than root,
        # then LCA must be in the left subtree
        # (because of BST property: left < root < right)
        if max(p.val, q.val) < root.val:
            return self.lowestCommonAncestor(root.left, p, q)
        
        # If BOTH p and q are greater than root,
        # then LCA must be in the right subtree
        elif min(p.val, q.val) > root.val:
            return self.lowestCommonAncestor(root.right, p, q)
        
        else:
            # We found the split point:
            # Either:
            # 1. p and q are on different sides of root
            # 2. One of them IS the root
            # This makes root the Lowest Common Ancestor
            return root